streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -465,6 +466,53 @@ public synchronized void onChange(final Thread thread,
         }
     }
 
+    final class DelegatingStateRestoreListener implements StateRestoreListener {
+        @Override
+        public void onRestoreStart(final TopicPartition topicPartition, final String storeName, final long startingOffset, final long endingOffset) {
```

Comment:
I wonder if it is worth refactoring this to remove duplication, i.e, add a Functional interface, then implement it three times to just perform the op on `globalStateRestoreListener` then have a method sth like:
```
performOnGlobalListener(GlobalListenerAction action, String storeName, TopicPartition partition) {
    if (globalStateRestoreListener != null) {
       try{    
         action.perform();
       } catch (final Exception e) {
       /// streams exception stuff
      }
   }
}
```
Other methods delegate to this new method. The guard and error handling are encapsulated in one place.  When we eventually get to java 8 we can just do a lambda call

Reply:
Hmm.. I either need to wrap all the parameters (note they are different for different functions) or use `String, TopicPartition, long..` (I do not know if it actually works at all).

I tried the first option and let me know what do you think, note it creates a new instance of the action on each call.. in terms of LOC it seems more even. Let me know WDYT:

```
interface GlobalListenerAction {
        void apply();
    }

    final class DelegatingStateRestoreListener implements StateRestoreListener {
        private void performOnGlobalListener(GlobalListenerAction action, String storeName, TopicPartition partition) {
            if (globalStateRestoreListener != null) {
                try{
                    action.apply();
                } catch (final Exception fatalUserException) {
                    throw new StreamsException(
                            String.format("Fatal user code error in store restore listener for store %s, partition %s.",
                                    storeName,
                                    partition),
                            fatalUserException);
                }
            }
        }

        @Override
        public void onRestoreStart(final TopicPartition topicPartition, final String storeName, final long startingOffset, final long endingOffset) {
            GlobalListenerAction onRestoreStartAction = new GlobalListenerAction() {
                @Override
                public void apply() {
                    globalStateRestoreListener.onRestoreStart(topicPartition, storeName, startingOffset, endingOffset);
                }
            };

            performOnGlobalListener(onRestoreStartAction, storeName, topicPartition);
        }

        @Override
        public void onBatchRestored(final TopicPartition topicPartition, final String storeName, final long batchEndOffset, final long numRestored) {
            GlobalListenerAction onBatchRestoredAction = new GlobalListenerAction() {
                @Override
                public void apply() {
                    globalStateRestoreListener.onBatchRestored(topicPartition, storeName, batchEndOffset, numRestored);
                }
            };

            performOnGlobalListener(onBatchRestoredAction, storeName, topicPartition);
        }

        @Override
        public void onRestoreEnd(final TopicPartition topicPartition, final String storeName, final long totalRestored) {
            GlobalListenerAction onRestoreEndAction = new GlobalListenerAction() {
                @Override
                public void apply() {
                    globalStateRestoreListener.onRestoreEnd(topicPartition, storeName, totalRestored);
                }
            };

            performOnGlobalListener(onRestoreEndAction, storeName, topicPartition);
        }
    }
```